source/dubregistry/internal/utils.d

```diff
@@ -1,15 +1,85 @@
 module dubregistry.internal.utils;
 
+import vibe.core.core;
+import vibe.core.concurrency;
+import vibe.core.task;
 import vibe.inet.url;
+import vibe.inet.path;
+
+import std.algorithm : any;
+import std.file : tempDir;
+import std.path;
+import std.process;
 
 URL black(URL url)
 {
-	if (url.username.length > 0) url.username = "***";
-	if (url.password.length > 0) url.password = "***";
+	if (url.username.length > 0)
+		url.username = "***";
+	if (url.password.length > 0)
+		url.password = "***";
 	return url;
 }
 
 string black(string url)
 {
 	return black(URL(url)).toString();
 }
+
+static immutable string logoOutputFolder = "uploads/logos";
+
+static immutable string[] logoFormats = [".webp", ".png"];
+
+/// 
+/// Params:
+///   file = the file to convert
+///   name = the logo name to put in uploads/logos
+///   deleteExisting = if false, throw an exception if the files already exist
+///   deleteFinish = if true, delete the input file after at least one successful conversion
+auto generateLogo(Path file, string name, bool deleteExisting = false, bool deleteFinish = true)
+{
+	if (!name.length)
+		throw new Exception("name may not be empty");
+	foreach (format; logoFormats)
+		if (existsFile(buildPath(logoOutputFolder, name ~ format)))
+		{
+			if (deleteExisting)
+				removeFile(buildPath(logoOutputFolder, name ~ format));
+			else
+				throw new Exception("logo " ~ format ~ " already exists");
+		}
+	auto t = runWorkerTaskH(&generateLogoUnsafe, Task.getThis(), file, name);
+	auto success = receiveOnlyCompat!(bool[logoFormats.length]);
```

Comment:
isn't Task.getThis() specific to this task then? We could simply wrap it all in a `runTask({ ... }).join()`

Reply:
Just tried sth. similar, vibe-d's Tasks cannot yet be joined from another thread, so you have to use message passing or a TaskCondition. Think your approach is fine, just that checking the package name in the explicit response should be more reliable when the code gets refactored at some later point.

Reply:
no I mean make the whole generateLogo function use runTask() so it gets a task ID and immediately after it join it in the same function so that the worker task would just send to that task. I think it would be the safest approach. Otherwise if two worker tasks run and one of them finishes early it might inject in there idk